manifests/config/item.pp

```diff
@@ -0,0 +1,104 @@
+# being a native type!
+#
+# @option name [String]
+#   A unique description of the desired configuration setting
+#
+# @param key
+#   The configuration key to be set
+#
+#     * You can get a list of all configuration keys by running:
+#       ``ldapsearch -H ldap://localhost:389 \
+#       -y /usr/share/puppet_ds389_config/<instance_name>_ds_pw.txt \
+#       -D "cn=SIMP Directory Manager" -s base -b "cn=config"``
+#
+# @param value
+#   The value that should be set for ``$key``
+#
+# @param admin_dn
+#   A DN with administrative rights to the directory
+#
+# @param pw_file
+#   A file containing the password for use with ``$admin_dn``
+#
+# @param service_name
+#   The Puppet resource name for the directory ``Service`` resource
+#
+# @param restart_service
+#   Whether or not to restart the directory server after applying this item
+#
+#     * There is a known list of items in the module data that will always
+#       generate a restart action
+#
+# @param host
+#   The host where the service is running
+#
+# @param port
+#   The port to which to connect
+#
+# @param base_dn
+#   The DN that holds the directory configuration items
+#
+# @param encrypt_connection
+#   If set to `false`, do not require encryption for off-host connections.
+#
+#   * Loopback connections are never encrypted.
+#
+define ds389::config::item (
+  String[1]                     $key,
+  String[1]                     $value,
+  String[2]                     $admin_dn,
+  Stdlib::Absolutepath          $pw_file,
+  Simplib::Systemd::ServiceName $service_name,
+  Boolean                       $restart_service    = false,
+  Simplib::Host                 $host               = '127.0.0.1',
+  Simplib::Port                 $port               = 389,
+  String[2]                     $base_dn            = 'cn=config',
+  Boolean                       $encrypt_connection = true
+) {
+
+  if stdlib::start_with($service_name, 'dirsrv@') {
+    $_service_name = $service_name
+  }
+  else {
+    $_service_name = "dirsrv@${service_name}"
+  }
+
+  if $host in ['0.0.0.0', '::'] {
+    $_host = '127.0.0.1'
+  }
+  else {
+    $_host = $host
+  }
+
+  $_ldap_command_base = "-x -D '${admin_dn}' -y '${pw_file}' -H ldap://${_host}:${port}"
+
+  if $_host in ['127.0.0.1', 'localhost', '::1'] {
+    $_ldap_command_extra = ''
+  }
+  elsif $encrypt_connection {
+    # Encrypt if going off system
+    $_ldap_command_extra = '-ZZ'
+  }
+
+  $_command = "echo -e \"dn: ${base_dn}\\nchangetype: modify\\nreplace: ${key}\\n${key}: ${value}\" | ldapmodify ${_ldap_command_extra} ${_ldap_command_base}"
+  $_unless = "test `ldapsearch ${_ldap_command_extra} ${_ldap_command_base} -LLL -s base -b '${base_dn}' '${key}' | grep -e '^${key}' | awk '{ print \$2 }'` == '${value}'"
+
+  # This should be a provider
+  exec { "Set ${base_dn},${key} on ${_service_name}":
+    command => $_command,
+    unless  => $_unless,
+    path    => ['/bin', '/usr/bin']
+  }
```

Comment:
A solution to the sensitive data problem mentioned in the header comments regarding `$values`:

* Make `$value` a [`Sensitive` data type](https://puppet.com/docs/puppet/6.19/lang_data_sensitive.html)  to redact mentions of `$value` from inside this code block.
* Use [`$value.node_encrypt::secret`](https://forge.puppet.com/binford2k/node_encrypt#usage) to encrypt it in the catalog until the agent uses it.

The drawback is that this usage of `node_encrypt::secret()` is for Puppet 6 only.  However, considering the [January EOL of PE2018.1/Puppet 5](https://puppet.com/docs/puppet-enterprise/product-support-lifecycle/), it would be reasonable policy where modules introduced with SIMP 6.6+ are not required to support Puppet < 6.17 ([as we did with modules introduced after SIMP 6.3 and Puppet < 5.5](https://github.com/simp/pupmod-simp-oath/commit/2a2840dec2035329a915d07aa7594a3c0ae558d4#diff-6ac3f79fc25d95cd1e3d51da53a4b21b939437392578a35ae8cd6d5366ca5485R13)).

Reply:
Exec can only take commands that are full `Sensitive` so casting it at the point where the parameter is assigned is most effective.

Hiera lookup-fu doesn't work because this is a defined type.

Reply:
Looking at `node_encrypt` again, it doesn't really have huge advantages in this case (or many cases if running in a normal puppet environment).

* `root` is the only user that can get to the compiled catalog
* `root` can also just go pluck these values out of the LDAP server configuration

I've been having some discussions with Ben Ford about just having an option to encrypt the entire catalog on disk where the language/server doesn't have to be involved and it looks like he's on board. This would "solve" the issue as well as keeping everything simple across the board.